src/core/bpf-firewall.c

```diff
@@ -50,6 +53,256 @@ enum {
         ACCESS_DENIED  = 2,
 };
 
+static void fix_jumps(struct bpf_insn *prog, uint length, uint offset) {
+        uint i;
+
+        /* Make all -1-length jumps go to the end of the block. */
+        for (i = 0; i < length; i++)
+                if ((prog[i].code & BPF_JMP) && (prog[i].off == -1))
+                        prog[i].off = offset - i;
+}
+
+static int add_port_lookup_instructions(
+                BPFProgram *p,
+                int map_fd,
+                int protocol,
+                bool is_ingress,
+                int verdict) {
+        /*
+         * We are interested only in destination port regardless of ingress or
+         * egress -- source port is typically ephemeral. TCP and UDP have the
+         * ports in the same place in header structure.
+         */
+        int tcp_port_offset = offsetof(struct tcphdr, dest);
```

Comment:
hmmm, given that it's 2017, we should really support both ipv4 and ipv6 from the start

Reply:
There already is IPv6 support (it's even tested :-). The port offset is added to the IP header length, which depends on the IP protocol family.